### ISSUE TITLE
Precedence of logical operators AND and OR

### ISSUE THREAD

**Issue by TigerHix** [PocketMine/PocketMine-MP#3266]
It seems to me that some uses of AND and OR are wrong.

For example in Entity class, in function **entityBaseTick($tickDiff)**, line 810:

    if(!$this->hasEffect(Effect::FIRE_RESISTANCE) and ($this->fireTicks % 20) === 0 or $tickDiff > 20){
        // do fire damage to entity
    }

The correct logic here would be:

> **IF**
> entity has fire resistance effect,
> **AND**
> it's time for entity to receive fire damage (fireTicks % 20 === 0), **OR** one second already passed after > the last execution of **entityBaseTick($tickDiff)** function.
> 
> **THEN**
> entity receives fire damage.

But the code actually interprets:

> **IF**
> entity has fire resistance effect **AND** it's time for entity to receive fire damage (fireTicks % 20 === 0),
> **OR**
> one second already passed after the last execution of **entityBaseTick($tickDiff)** function.
> 
> **THEN**
> entity receives fire damage.

Because AND has a higher precedence than OR, therefore
    if(!$this->hasEffect(Effect::FIRE_RESISTANCE) and ($this->fireTicks % 20) === 0 or $tickDiff > 20){
is equivalent to
    if((!$this->hasEffect(Effect::FIRE_RESISTANCE) and ($this->fireTicks % 20) === 0) or $tickDiff > 20){ // This is the wrong logic!
but not
    if(!$this->hasEffect(Effect::FIRE_RESISTANCE) and (($this->fireTicks % 20) === 0 or $tickDiff > 20)){ // This is the correct logic!

Another example, also in Entity class, in function **updateMovement()**, line 846:

    if($diffPosition > 0.04 or $diffRotation > 2.25 and ($diffMotion > 0.0001 and $this->getMotion()->lengthSquared() <= 0.00001)){
        // update entity movement
    }

The correct logic here would be:

> **IF**
> difference in position **OR** difference in rotation is enough for PocketMine to send the updated movements,
> **AND**
> entity has a motion.
> 
> **THEN**
> update entity movement.

While the code interprets:

> **IF**
> difference in position
> **OR** 
> difference in rotation is enough for PocketMine to send the updated movements **AND** entity has a motion.
> 
> **THEN**
> update entity movement.

Same logical fault like the fire damage one. The AND and OR operators are incorrectly used it seemed. The precedences are inverted.

There is also one at line 857:

    if($diffMotion > 0.0025 or ($diffMotion > 0.0001 and $this->getMotion()->lengthSquared() <= 0.0001)){

The brackets are unnecessary at all, **unless** the code author thinks OR has a higher precedence than AND.

*TL;DR - the code is missing brackets.*

**Comment by PEMapModder** [comment]
Make pull requests?

**Comment by TigerHix** [comment]
:S Well after all I'm not so sure about this. Hoping to get confirmation from the team so I could have a PR of fixing possible mistakes that I had found.
